README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# nsrr

<!-- badges: start -->

[![Travis build
status](https://travis-ci.com/muschellij2/nsrr.svg?branch=master)](https://travis-ci.com/muschellij2/nsrr)
[![AppVeyor Build
Status](https://ci.appveyor.com/api/projects/status/github/muschellij2/nsrr?branch=master&svg=true)](https://ci.appveyor.com/project/muschellij2/nsrr)
[![Codecov test
coverage](https://codecov.io/gh/muschellij2/nsrr/branch/master/graph/badge.svg)](https://codecov.io/gh/muschellij2/nsrr?branch=master)
<!-- badges: end -->

The goal of nsrr is to allow users to access data from the National
Sleep Research Resource (’NSRR’) (<https://sleepdata.org/>) through an
`R` interface.

Why `R`? Many packages in `R` can read and process accelerometry, such
as the [`GGIR`](https://cran.r-project.org/package=GGIR),
[`ActivityIndex`](https://cran.r-project.org/package=ActivityIndex), and
[`accelerometry`](https://cran.r-project.org/package=accelerometry);
other packages such as
[`edfReader`](https://cran.r-project.org/package=edfReader) can read in
EDF data for polysomnography data. Also, the
[`xml2`](https://cran.r-project.org/package=xml2) package can easily
read in XML annotations into `R`. We believe the interplay with these
packages, amongst others, allow for a flexible framework to download,
process, and visualize data. The `nsrr` package is simply the entry
point into navigating the files available and downloading the data.

## Installation

You can install the released version of nsrr from
[CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("nsrr")
```

## Token

To retrieve your NSRR token, go to <https://sleepdata.org/dashboard>,
and sign in. This token will allow you access to any data sets you have
requested access to. If you do not have access, then it will allow you
to download files that are publicly available.

Set the token by adding this to your `~/.Renviron` file:

``` r
NSRR_TOKEN="YOUR TOKEN GOES HERE"
```

The token is accessible via `token = Sys.getenv("NSRR_TOKEN")`. Each
`nsrr` function also has the argument `token` to pass through if you do
not wish to set it.

To determine if you are authenticated, you can use:

``` r
library(nsrr)
nsrr_auth()
$authenticated
[1] TRUE

$username
[1] "muschellij2"

$full_name
[1] ""

$first_name
NULL

$last_name
NULL

$email
[1] "muschellij2@gmail.com"
```

## Examples

### NSRR data sets

Here is how you can access the NSRR datasets list:

``` r
library(nsrr)
df = nsrr_datasets()
DT::datatable(df)
```

<img src="man/figures/README-example-1.png" width="100%" />

### NSRR data set files

Here we first get a list of the files in the `datasets` sub-directory
from the `shhs` data set:

``` r
df = nsrr_dataset_files("shhs", path = "datasets")
head(df)
  dataset                         full_path    folder
1    shhs                  datasets/archive datasets/
2    shhs             datasets/hrv-analysis datasets/
3    shhs             datasets/CHANGELOG.md datasets/
4    shhs           datasets/KNOWNISSUES.md datasets/
5    shhs datasets/shhs1-dataset-0.15.0.csv datasets/
6    shhs datasets/shhs2-dataset-0.15.0.csv datasets/
                 file_name is_file file_size
1                  archive   FALSE         0
2             hrv-analysis   FALSE         0
3             CHANGELOG.md    TRUE     11010
4           KNOWNISSUES.md    TRUE     11421
5 shhs1-dataset-0.15.0.csv    TRUE  24322962
6 shhs2-dataset-0.15.0.csv    TRUE  11897377
                 file_checksum_md5 archived
1                             <NA>    FALSE
2                             <NA>    FALSE
3 69bb54a32cdfc7bdddc13276b7c858c1    FALSE
4 8598129123baa60e16977dc24aa780af    FALSE
5 3f26d37ec97e2bc88776850a31715398    FALSE
6 21807d854010f036fd0d4f006eeed49d    FALSE
```

### Downloading NSRR data set files

We can then download the `CHANGELOG.md` file as it’s publicly
accessible.

``` r
url = nsrr_download_url("shhs", path = "datasets/CHANGELOG.md")
# print URL
dl = nsrr_download_file("shhs", path = "datasets/CHANGELOG.md")
dl$outfile
[1] "/var/folders/1s/wrtqcpxn685_zk570bnx9_rr0000gr/T//RtmpJUlkmk/filea17e12cd407.md"
cat(head(readLines(dl$outfile)), sep = "\n")
## 0.15.0 (November 18, 2019)

- Remove EEG spectral summary variables
- Add notes to height and weight varibles about top and bottom coding
- The CSV datasets generated from a SAS export are located here:
  - `\\rfawin\bwh-sleepepi-shhs\nsrr-prep\_releases\0.15.0\`
```

### Listing All NSRR data set files

To list all the files, recursively, you would run:

``` r
nsrr_all_dataset_files("shhs")
```

but it may take some time.

### EDF files

The [`edfReader`](https://cran.r-project.org/package=edfReader) can read
in EDF fields for polysomnography data. **Work in Progress - need access
to EDF data**.

``` r
files = nsrr_dataset_files("shhs", path= "polysomnography/edfs/shhs1")
dl = nsrr_download_file("shhs", path = "polysomnography/edfs/shhs1/shhs1-200001.edf")
```


NEWS.md
# nsrr 0.2.1

- Added `check_filesize` to speed up downloads.

# nsrr 0.2.0

- Major refactor, especially with SSL certificate problems
- Next release to CRAN.

# nsrr 0.1.4

- Minor fixes to vignette due to `pandoc`. 
- Added a few more tests to increase code coverage.

# nsrr 0.1.3

- Now `nsrr_datasets` returns all data sets, not just the first page, fixes #2.

# nsrr 0.1.2

- Fixing examples.

# nsrr 0.1.0

* First submission to CRAN. 
* Added a `NEWS.md` file to track changes to the package.

* Added download functionality to the package.



tests/testthat/test-token.R
testthat::context(desc = "Testing data sets")


testthat::test_that(
  "Testing if data sets returns a data.frame if not authenticated", {

    df = nsrr_datasets(token = NULL)

    if (all(attributes(df)$status_code == 200)) {
      testthat::expect_is(df, "data.frame")
    }

  })

testthat::test_that("Testing if data sets returns a data.frame", {

  token = nsrr_token()
  if (!is.null(token)) {
    df = nsrr_datasets(token = token)
    if (all(attributes(df)$status_code == 200)) {
      testthat::expect_is(df, "data.frame")
    }
  }

})


R/nsrr_token.R
#' NSRR Token
#'
#' @param token Token for NSRR resources.  Found at
#' \url{https://sleepdata.org/token}
#'
#' @return A character vector or NULL
#' @export
#'
#' @examples
#' is.null(nsrr_token())
#' if (!is.null(nsrr_token())) {
#'    res = nsrr_auth()
#'    res$authenticated
#' }
#' bad_res = nsrr_auth("")
#' bad_res$authenticated
#' nsrr_have_token()
#' nsrr_authenticated()
nsrr_token = function(token = NULL) {
  if (is.null(token)) {
    token = Sys.getenv("NSRR_TOKEN")
  }
  if (length(token) > 0) {
    if (token == "") {
      token = NULL
    }
  }
  token = trimws(token)
  if (length(token) == 0) {
    token = NULL
  }
  token
}

#' @export
#' @rdname nsrr_token
nsrr_have_token = function(token = NULL) {
  !is.null(nsrr_token(token = token))
}

#' @export
#' @rdname nsrr_token
nsrr_auth = function(token = NULL) {

  res = nsrr_api(
    path = "/account/profile.json",
    query = list(),
    token = token)
  result = httr::content(res)
  result
}

#' @export
#' @rdname nsrr_token
nsrr_authenticated = function(token = NULL) {
  auth = nsrr_auth(token = token)
  auth$authenticated
}



R/nsrr_version.R
#' NSRR Gem version online
#'
#' @return A character of the version
#' @export
#' @importFrom utils packageVersion
#'
#'
#' @examples
#' nsrr_version()
#' nsrr_gem_version()
#'
nsrr_version = function() {
  # "0.4.0"
  ver = utils::packageVersion("nsrr")
  as.character(ver)
}

#' @export
#' @rdname nsrr_version
#' @importFrom stats na.omit
nsrr_gem_version = function() {
  url = "https://raw.githubusercontent.com/nsrr/nsrr-gem/master/lib/nsrr/version.rb"
  res = httr::GET(url)
  httr::stop_for_status(res)
  cr = httr::content(res)
  ss = strsplit(cr, "\n")[[1]]
  ss = trimws(ss)
  ords = c("MAJOR","MINOR","TINY","BUILD")
  search_string = paste0("^(", paste(ords, collapse = "|"), ")")
  ss = ss[ grepl(search_string, ss)]
  ss = sub("nil", "", ss)
  ss =  sub("#.*", "", ss)
  ss = strsplit(ss, " ")
  n_ss = sapply(ss, function(x) {
    xx = x[grepl(search_string, x)]
    stopifnot(length(xx) <= 1)
    xx
  })
  names(ss) = n_ss
  ss = sapply(ss, function(x) {
    xx = x[!grepl(search_string, x)]
    xx = xx[ xx != "="]
    xx = xx[ xx != ""]
    if (length(xx) == 0) {
      xx = ""
    }
    xx = paste0(xx, collapse = ".")
    xx
  })
  ss
  ss = sub(" = ", "=", ss)
  m = match(names(ss), ords)
  m = stats::na.omit(m)
  ss = ss[ m ]
  ss = paste(ss, collapse = ".")
  ss = sub("[.]$", "", ss)
  ss = gsub("[..]", ".", ss)
  ss
}


#' @export
#' @rdname nsrr_version
nsrr_api_version = function() {
  "v1"
}


R/nsrr_website.R
#' The NSRR Website URL
#'
#' @return A character value of the website
#' @export
#'
#' @examples
#' nsrr_website()
nsrr_website = function() {
  website = "https://sleepdata.org"
  website
}

#' @export
#' @rdname nsrr_website
nsrr_api_url = function() {
  website = nsrr_website()
  url = paste0(website, "/api/", nsrr_api_version())
  url
}

#' NSRR API Functionality
#'
#' @param path path to endpoint for NSRR API
#' @param token Authorization Token for NSRR, passed to
#' \code{\link{nsrr_token}}
#' @param query query to pass to \code{\link{GET}}
#' @param ... additional arguments to pass to \code{\link{GET}}
#'
#' @return A response object
#' @export
nsrr_api = function(path, query = list(), token = NULL,
                    ...) {
  token = nsrr_token(token = token)

  url = paste0(nsrr_api_url(), path)

  query$auth_token = token
  res = httr::with_config(
    config = httr::config(ssl_verifypeer = FALSE),
    httr::GET(url, query = query, ...)
  )
  res
}



R/nsrr_download.R
#' NSRR Download file
#'
#' @inheritParams nsrr_datasets
#' @param path full path to the file.
#'
#' @return A \code{data.frame} of the data sets and their endpoints
#' @export
#' @importFrom httr progress write_disk
#' @importFrom tools file_ext
#'
#' @examples
#' if (requireNamespace("xml2", quietly = TRUE) &
#'     requireNamespace("rvest", quietly = TRUE)) {
#'   res = httr::with_config(
#'     config = httr::config(ssl_verifypeer = FALSE),
#'     httr::GET("https://sleepdata.org/datasets/shhs/files/datasets")
#'   )
#'   if (httr::status_code(res) < 400) {
#'     doc = httr::content(res)
#'     tab = rvest::html_table(doc)[[1]]
#'     path = tab$X2
#'     path = path[ grepl("shhs-data-dictionary-.*-domains", path)]
#'     path = path[1]
#'     path = paste0("datasets/", path)
#'   } else {
#'     path = "datasets/shhs-data-dictionary-0.16.0-domains.csv"
#'   }
#' } else {
#'   path = "datasets/shhs-data-dictionary-0.16.0-domains.csv"
#' }
#' dataset = "shhs"
#' nsrr_download_url(dataset, path, token = "")
#' if (nsrr_have_token()) {
#'   res = nsrr_download_file(dataset, path, check_md5 = FALSE,
#'                            check_filesize = FALSE)
#'   testthat::expect_true(res$success)
#' }
#' url = nsrr_download_url("shhs", path = "datasets/CHANGELOG.md",
#'                         token = NULL)
#' res = nsrr_download_file("shhs", path = "datasets/CHANGELOG.md",
#'                          token = NULL)
#' testthat::expect_true(res$correct_md5)
#' \donttest{
#'   res = nsrr_download_file("shhs", path = "datasets/CHANGELOG.md",
#'                            token = NULL, check_md5 = FALSE)
#'   testthat::expect_null(res$correct_md5)
#'   if (nsrr_have_token()) {
#'     path = "biostatistics-with-r/shhs1.txt"
#'     res = nsrr_download_file(dataset, path)
#'   }
#' }
nsrr_download_url = function(
  dataset,
  path,
  token = nsrr_token()
) {
  # stopifnot(!is.null(token))
  # stopifnot(length(token) == 1)
  ver = nsrr_version()

  if (!is.null(token)) {
    if (token == "") {
      token = NULL
    }
  }
  fname = paste(
    nsrr_website(),
    "datasets",
    dataset,
    "files", sep = "/")
  if (!is.null(token)) {
    fname = paste(fname, paste0("a/", token), sep = "/")
  }
  fname = paste(
    fname,
    "m",
    paste0("nsrr-r-v", gsub("[.]", "-", ver)),
    # paste0("nsrr-gem-v", gsub("[.]", "-", ver)),
    path, sep = "/")
  return(fname)
}

#' @export
#' @rdname nsrr_download_url
#' @param check_md5 check if MD5 checksum agrees when downloaded
#' @param check_filesize check the file size versus the `API`
#' @importFrom digest digest
nsrr_download_file = function(
  dataset, path,
  token = nsrr_token(),
  check_md5 = TRUE,
  check_filesize = TRUE
) {
  stopifnot(length(path) == 1)
  url = nsrr_download_url(dataset, path, token = token)
  if (check_filesize || check_md5) {
    file_size = nsrr_dataset_files(dataset, path = path, token = token)
    file_md5 = file_size$file_checksum_md5
    file_size = file_size$file_size
  }
  ext = tools::file_ext(url)
  tfile = tempfile(fileext = paste0(".", ext))
  query = list()
  query$auth_token = token
  res = httr::with_config(
    config = httr::config(ssl_verifypeer = FALSE),
    httr::GET(
      url,
      if (interactive()) httr::progress(),
      httr::write_disk(path = tfile),
      query = query
    )
  )
  if (!check_filesize) {
    correct_size = TRUE
  } else {
    raw_content = httr::content(res, as = "raw")
    size <- length(raw_content)
    correct_size = size == file_size
  }
  if (check_md5) {
    md5 = digest::digest(tfile, file = TRUE, algo = "md5")
    correct_md5 = md5 == file_md5
    if (length(correct_md5) == 0) {
      warnings("MD5 file size not found, reporting TRUE")
      correct_md5 = TRUE
    }
  } else {
    correct_md5 = TRUE
  }


  success = httr::status_code(res) < 400 &
    correct_md5 & correct_size
  if (!success) {
    warning(paste0("Not successful download - ",
                   "likely not authorized for resource"))
  }
  L = list(response = res,
           outfile = tfile,
           correct_size = correct_size,
           success = success
  )
  if (check_md5) {
    L$correct_md5 = correct_md5
  }
  return(L)
}


R/nsrr_all_files.R
#' NSRR All Dataset files
#'
#' @inheritParams nsrr_datasets
#'
#' @param max_files maximum files to return if not wanting to go through
#' all folders recursively.
#' @return A \code{data.frame} of the data sets and their endpoints
#' @export
#'
#' @importFrom httr GET content
#' @importFrom jsonlite fromJSON
#'
#' @examples
#' dataset = "shhs"
#' df = nsrr_all_dataset_files(dataset, max_files = 4)
#' testthat::expect_lte(nrow(df), 20)
#' # keep as donttest - takes a long time
#' \dontrun{
#' df = nsrr_all_dataset_files(dataset)
#' }
#'
nsrr_all_dataset_files = function(dataset, token = nsrr_token(),
                                  max_files = Inf) {
  df = nsrr_dataset_files(dataset, path = NULL, token = token)
  df$is_folder = !df$is_file
  while (any(df$is_folder) & nrow(df) <= max_files) {
    ind = which(df$is_folder)[1]
    ipath = df$full_path[ind]
    print(ipath)
    df = df[-ind,]
    res = nsrr_dataset_files(dataset, path = ipath, token = token)
    res$is_folder = !res$is_file
    df = rbind(df, res)
  }
  return(df)
}


R/nsrr_datasets.R

#' NSRR data sets
#'
#' @inheritParams nsrr_token
#' @param page which page to grab.  Increment over successive requests to
#' retrieve all datasets. A request that return NULL or a number of datasets
#' less than 18 indicates the last page.
#'
#' @return A \code{data.frame} of the data sets and their endpoints
#' @export
#'
#' @importFrom httr GET content
#' @importFrom jsonlite fromJSON
#'
#' @examples
#' df = nsrr_datasets(page = 1)
#' on_cran = !identical(Sys.getenv("NOT_CRAN"), "true")
#' on_ci <- nzchar(Sys.getenv("CI"))
#' local_run = grepl("musch", tolower(Sys.info()[["user"]]))
#' run_example = !on_cran || on_ci || local_run
#' if (run_example) {
#'   df = nsrr_datasets()
#'   if (attributes(df)$status_code == 200) {
#'     testthat::expect_is(df, "data.frame")
#'     slugs = c("abc", "bestair", "chat", "ccshs", "cfs",
#'               "heartbeat", "hchs", "homepap", "haassa", "learn")
#'     testthat::expect_true(all(slugs %in% df$slug))
#'   }
#' }
nsrr_datasets = function(token = nsrr_token(),
                         page = NULL) {
  # website = nsrr_api_url()
  # datasets = paste0(website, "/datasets.json")
  max_pages = 100
  check_max = TRUE
  if (is.null(page)) {
    pages = 1:max_pages
  } else {
    pages = page
    check_max = FALSE
  }
  df = vector(mode = "list", length = length(pages))
  for (ipage in seq_along(pages)) {
    page = pages[ipage]
    query = list()
    query$page = page
    res = nsrr_api(
      path = "/datasets.json",
      query = query,
      token = token)
    if (httr::status_code(res) == 500) {
      warning(
        paste0(
          "Please contact the NSRR Team - the server",
          " has indicated an error, returning NULL"
          ))
      return(NULL)
    }
    # res = httr::GET(datasets, query = query)
    httr::stop_for_status(res)
    x = httr::content(res, as = "text")
    x = jsonlite::fromJSON(x, flatten = TRUE)
    if (NROW(x) > 0) {
      x$slug = sub("/datasets/", "", trimws(x$path))
      x$slug = sub(".json$", "", x$slug)
      x$files = sub(".json", "", x$path)
      x$files = paste0(x$files, "/files.json")
      x$status_code = httr::status_code(res)
      x$page = ipage
      df[[ipage]] = x
    } else {
      break
    }
  }
  if (check_max && ipage >= max_pages) {
    warning(
      paste0(
        "May not have received all data - pass in page",
        " argument to paginate through")
    )
  }
  x = do.call("rbind", df)
  attr(x, "status_code") = unique(x$status_code)
  x$status_code = NULL

  return(x)
}

#' @export
#' @rdname nsrr_datasets
#' @examples
#' nsrr_dataset(dataset = "shhs", token = "")
nsrr_dataset = function(
  dataset = NULL,
  token = nsrr_token()) {

  res = nsrr_api(
    path = paste0("/datasets/", dataset, ".json"),
    token = token)

  x = httr::content(res, as = "text")
  x = jsonlite::fromJSON(x, flatten = TRUE)
  attr(x, "status_code") = httr::status_code(res)
  return(x)
}

#' @export
#' @rdname nsrr_datasets
#' @param dataset a dataset \code{"slug"}, one from
#' \code{\link{nsrr_datasets}}
#' @param path a folder or file path inside the dataset
#' @examples
#' dataset = "shhs"
#' token = NULL
#' df = nsrr_dataset_files(dataset)
#' ddf = nsrr_dataset_files(dataset, path = df$full_path[1])
#'
#'
#' dataset = "shhs"
#' token = NULL
#' df = nsrr_dataset_files(dataset)
#' nsrr_dataset_files(dataset = "wecare")
#'
#' testthat::expect_error(nsrr_dataset_files(), "one data")
#' testthat::expect_error(nsrr_dataset_files(c("shhs", "chat")), "one data")
nsrr_dataset_files = function(
  dataset = NULL,
  path = NULL,
  token = nsrr_token()) {
  msg = "Need to specify one data set"
  if (is.null(dataset)) {
    stop(msg)
  }
  if (length(dataset) > 1) {
    stop(msg)
  }
  df = nsrr_datasets(token = token)
  if (!dataset %in% df$slug) {
    warning("Dataset not in set from NSRR")
    run_path = paste0("/datasets/", dataset, "/files.json")
  } else {
    idf = df[ df$slug %in% dataset, ]
    run_path = idf$files
  }
  query = list()
  query$path = path
  res = nsrr_api(
    path = run_path,
    query = query,
    token = token)

  httr::stop_for_status(res)
  cr = httr::content(res)
  x = httr::content(res, as = "text")
  if (is.null(cr) ||
      (is.character(x) && all(x == ""))) {
    warning("Content was NULL, returning the response for debugging")
    return(res)
  }

  x = jsonlite::fromJSON(x, flatten = TRUE)
  attr(x, "status_code") = httr::status_code(res)
  return(x)
}


